core/src/main/java/org/eclipse/hono/util/MessageHelper.java

```diff
@@ -835,4 +838,24 @@ public static Message getShallowCopy(final Message message) {
     public static Object getCorrelationId(final Message message) {
         return Optional.ofNullable(message.getCorrelationId()).orElse(message.getMessageId());
     }
+
+    /**
+     * Gets the string value of a <em>message-id</em> type property value.
+     *
+     * @param messageIdValue The value of a property of type <em>message-id</em>.
+     * @return The string value.
+     * @throws IllegalArgumentException if the value has an unsupported type.
+     */
+    public static String getMessageIdStringValue(final Object messageIdValue) {
+        Objects.requireNonNull(messageIdValue);
+
+        if (messageIdValue instanceof String || messageIdValue instanceof UnsignedLong
+                || messageIdValue instanceof UUID) {
+            return messageIdValue.toString();
+        } else if (messageIdValue instanceof Binary) {
+            return Base64.getEncoder().encodeToString(((Binary) messageIdValue).getArray());
+        } else {
```

Comment:
I do not think that this will work because it effectively changes the type of the correlation ID to `String` before sending it to the device. The device will then include this (String typed) correlation ID in its response sent to the adapter and the adapter will put it into the response AMQP message as a String typed correlation ID.
Thus, when an application sends a request/response command and uses e.g. a UUID as the correlation ID, I expect it to fail in comparing the response's correlation ID with the expected value ...

Reply:
The idea here was that we explicitly document that the command response correlation ID will be a String:

> Response Message Format
> correlation-id | yes | properties | string
> MUST contain the correlation ID used to match the command message with the response message containing the result of execution on the device. The ID here is the string value of the correlation ID set on the command message (a binary value is base64-encoded here).

If an application chooses to use a non-String correlation ID for a command message, it will have to match the command response according to the string representation.

I see that putting such a requirement on the application is very much similar to requiring it to use a command message with a string correlation ID in the first place. 

Still I would rather prefer the above all-types-allowed solution, not rejecting command messages with a non-string correlation id. How correlation ids are to be matched is not described in the AMQP spec anyway.

Reply:
I have to disagree. FMPOV the restriction that we currently have in place, leaves no doubt as to what a client needs to put in there and how IDs are to be matched. Receiving back the String representation of the ID that a client put into the request in the response seems very strange to me.
FMPOV we either properly support all types, i.e. we also encode the message-id type into the reply-id or we stick to just supporting string typed message-ids.

Reply:
Well, I don't have a strong preference here, requiring a string correlation id in the command message is fine with me as well. I've amended the commit accordingly.